xla/service/collective_quantizer.cc

```diff
@@ -279,45 +320,55 @@ absl::Status MatchQuantization(HloInstruction* instr, bool* changed) {
     subgraph = IsSupportedQuantization(instr->users()[0]);
   }
 
-  if (subgraph.has_value()) {
-    HloInstruction* coll_operand = instr->mutable_operand(0);
-
-    HloInstruction *new_binary, *new_clamp;
-    // When there is a quantization, insert the scale and clamp ops.
-    if (subgraph->binary) {
-      HloInstruction* new_scale_bcast = instr->AddInstruction(
-          subgraph->scale_bcast->CloneWithNewShape(coll_operand->shape()));
-      new_binary = instr->AddInstruction(subgraph->binary->CloneWithNewOperands(
-          coll_operand->shape(), {coll_operand, new_scale_bcast}));
-      HloInstruction* new_clamp_lower =
-          instr->AddInstruction(subgraph->clamp->operand(0)->CloneWithNewShape(
-              coll_operand->shape()));
-      HloInstruction* new_clamp_upper =
-          instr->AddInstruction(subgraph->clamp->operand(2)->CloneWithNewShape(
-              coll_operand->shape()));
-      new_clamp = instr->AddInstruction(subgraph->clamp->CloneWithNewOperands(
-          coll_operand->shape(),
-          {new_clamp_lower, new_binary, new_clamp_upper}));
+  if (!subgraph.has_value()) {
+    return absl::OkStatus();
+  }
+
+  if (subgraph->scale_bcast) {
+    // The scale must be replicated across all devices.
+    TF_ASSIGN_OR_RETURN(
+        bool scale_is_replicated,
+        InstrIsReplicated(instr->parent()->parent(), subgraph->scale_bcast));
```

Comment:
you also should check the clamp values are replicated. Or for simplicity, check they are constant.